Optimize timings around k_sleep

nordic-krch opened the pull request:
In embedded devices it's often costly to access peripheral registers (compared to accessing RAM memory). For example in nrf54h20 which runs at 320 MHz reading register in a peripheral takes 160 cycles and system clock counter has 64 bits which requires 2 reads. It is that high because peripherals are in low power domain which runs at much lower frequency. It should be avoided if possible. The peripheral that is the most often accessed is the system clock. In case of `k_sleep` system clock is accessed multiple times since call returns number of ticks left (valid only if thead is waken up prematurely). Current implemention is doing following:
- read the system clock to calculate expected expiration tick
- sets timeout
- swaps the context
- read the system clock to compare it against expected expiration tick

Those reads are redundant and PR proposes following improvements:
- Extend `z_add_timeout` to return tick when timeout will expire. It is calculated anyway so there is no additional cost. It removes the need for reading the system clock to calculate expected expiration tick
- Add internal API to detect if timer was aborted and use that to determine if thread was waken up prematurely so that system clock read is needed only in case of premature wake up.

Those changes removes 2 reads.
Following changes has almost no impact on code size (+28 bytes on qemu_cortex_m3, +22 bytes on qemu_riscv32) but significantly reduces time spent in `z_tick_sleep`. I've measured time spent in that function excluding `z_add_timeout` and `z_swap` and got following results:

| target     | before | after |
| ----------- | -------- | ------ | 
| nrf52      | 7.7 us | 3.5 us |
| nrf54l15 | 4us | 1.3 us |
| nrf54h20| 4 us | 1.5 us| 
| stm32    | 3.6 us| 1.6 us|

In all cases there is 2-3 times reduction.

It reduces core active time in case when there are short periodic operations:

```
while (1) {
  k_msleep(1);
  // short op
}
```

nordic-krch committed:
kernel: sched: sleep: Use value returned by z_add_timeout

z_tick_sleep function needs to calculate expected wakeup tick. It
required reading current system clock tick. It can be costly since
it is a register access. When timeout is added this value is
calculated and z_add_timeout returns it. It can be used instead to
optimize sleep performance.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

npitre reviewed (commented):
Minor comments:

- There is a spurious blank line added to `z_add_timeout()`.

- Do you really need to cast `z_add_timeout()` to void? IMHO this is adding
  unneeded distraction to the code. May be warranted when a function returns
  an error code that you explicitly mean to ignore, but that's not the case
  here.

peter-mitsis commented on kernel/include/timeout_q.h:
```diff
@@ -37,6 +37,14 @@ static inline bool z_is_inactive_timeout(const struct _timeout *to)
 
+static inline bool z_is_expired_timeout(const struct _timeout *to)
+{
+	/* When timeout is aborted then dticks is not reset. It can be used to detect
+	 * if timeout was aborted or expired.
+	 */
```
Although it is true that when a timeout is aborted _dticks_ is not reset, the value of _dticks_ is insufficient to determine if the timeout was aborted or expired.

Consider the following scenario ... _sys_clock_announce()_ is called and there are multiple timeouts ready to expire. That is, there are multiple timeouts whose _dticks_ field is set to 0. The handlers for each of these timeouts are called as each timeout is processed. Should one of the earlier timeouts abort one of the later timeouts that are ready to expire, then that later timeout will get aborted while its _dticks_ is 0 and its handler will not execute.

peter-mitsis commented on kernel/include/timeout_q.h:
```diff
@@ -52,9 +51,9 @@ static inline void z_init_thread_timeout(struct _thread_base *thread_base)
 extern void z_thread_timeout(struct _timeout *timeout);
 
-static inline void z_add_thread_timeout(struct k_thread *thread, k_timeout_t ticks)
+static inline k_ticks_t z_add_thread_timeout(struct k_thread *thread, k_timeout_t ticks)
```
Since this is now returning a value with meaning, I think that it is worthwhile to note what this return value means (system tick when timeout is expected to expire).

peter-mitsis commented on kernel/timeout.c:
```diff
@@ -100,11 +100,12 @@ static int32_t next_timeout(void)
 
-void z_add_timeout(struct _timeout *to, _timeout_func_t fn,
-		   k_timeout_t timeout)
+k_ticks_t z_add_timeout(struct _timeout *to, _timeout_func_t fn, k_timeout_t timeout)
```
As with z_add_thread_timeout(), I think it is worthwhile to note what this return value means (system tick when timeout is expected to expire).

peter-mitsis reviewed (commented):
This is mostly okay--just a couple of things to polish.

andyross reviewed (commented):
Generally concur with the other comments.  The timeout logic gets pretty hairy, so there are many edge cases to worry about.

peter-mitsis commented on kernel/include/timeout_q.h:
```diff
@@ -63,14 +81,16 @@ k_ticks_t z_timeout_remaining(const struct _timeout *timeout);
 /* Stubs when !CONFIG_SYS_CLOCK_EXISTS */
 #define z_init_thread_timeout(thread_base) do {} while (false)
 #define z_abort_thread_timeout(to) do {} while (false)
+#define z_is_aborted_timeout(th) false
```
Missing `#define z_is_aborted_thread_timeout(to) false`

nordic-krch committed:
kernel: sched: Optimize sleeping function

Accessing system timer registers can be costly and it shall be avoided
if possible. When thread is waken up in z_tick_sleep it may be because
timeout expired or because thread was waken up before sleeping period
passed.

Add function to detect if timeout is aborted (before it was expired).
Use it in the sleep function and avoid reading system ticks if timeout
was not aborted.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
kernel: timeout: z_add_timeout to return expiration tick

Add return value to z_add_timeout. It returns system tick when timeout
will expire.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on kernel/sched.c:
```diff
@@ -1091,23 +1091,24 @@ static int32_t z_tick_sleep(k_timeout_t timeout)
 	(void)z_swap(&_sched_spinlock, key);
 
+	/* Reading system clock might be costly (register access) so avoid it if possible.
+	 * If thread timeout is expired then whole sleep time passed.
```
Removed.

nordic-krch commented:
@peter-mitsis @andyross can you take another look?
